src/catchup/CatchupWork.h

```diff
@@ -14,31 +14,44 @@
 namespace stellar
 {
 
-class LedgerRange;
-class CheckpointRange;
 class HistoryManager;
 class Bucket;
 class TmpDir;
+struct LedgerRange;
+struct CheckpointRange;
 
 // Range required to do a catchup.
 //
-// If second is true, this catchup requires downloading and applying buckets
-// for the first.first() and then downloading and applying transactions.
+// For initial catchup (after new-db) we have lastClosedLedger ==
+// LedgerManager.GENESIS_LEDGER_SEQ. In that case CatchupConfiguration::count()
+// and CatchupConfiguration::toLedger() are taken into consideration. Depending
+// on all of those values only one of "apply buckets" and "apply transactions"
+// can be executed, or both of them. The values are calculated in such a way
+// that transactions from at least count() ledgers are available in txhistory
+// table.
 //
-// If second is false, this catchup requires downloading and applying
-// transactions for whole range of ledgers.
+// If mApplyBuckets is true, this catchup requires downloading and applying
```

Comment:
For current implementation:
- `mApplyLedgers` is a range of ledgers to apply, it does not depent on `mApplyBuckets` (the comment was wrong)
- bucket are always applied at `mApplyLedgers.mFirst - 1`
- it is not possible to have buckets-only catchup, always at least one ledger is applied as well (the code was too complicated without that condition and the gains were minimal

I've updated the comment.

Reply:
> mApplyLedgers is a range of ledgers to apply, it does not depent on mApplyBuckets (the comment was wrong)

OK cool

> it is not possible to have buckets-only catchup, always at least one ledger is applied as well (the code was too complicated without that condition and the gains were minimal

The rest of my comment is still valid: if instead of doing what you're doing you were allowing to have an empty interval for the range of transactions to apply, you would end up with the same code simplification (reason the code was complicated before was that it made the range ~dynamic) *and* avoid the regression introduced in this PR.

The regression introduced:
now because we always replay at least one set of transactions, running something like `./stellar-core catchup 127/0` ends up applying buckets from the previous checkpoint and replays 64 ledgers (so it performs quite a few more roundtrips with the archive subsystem for no good reason at all).